### PR TITLE
Remap instrument-coverage line numbers in doctests

### PR DESCRIPTION
**Pull request by Swatinem**
This uses the `SourceMap::doctest_offset_line` method to re-map line
numbers from doctests. Remapping columns is not yet done, and rustdoc
still does not output the correct filename when running doctests in a
workspace.

Part of #79417 although I dont consider that fixed until both filenames
and columns are mapped correctly.

r? @richkadel 

I might jump on zulip the comming days. Still need to figure out how to properly write tests for this, and deal with other doctest issues in the meantime.

### CODE REVIEW

File: compiler/rustc_mir/src/transform/coverage/mod.rs
Comment (richkadel) [inline compiler/rustc_mir/src/transform/coverage/mod.rs:494]: I don't see a reason to move these two assignments into `make_code_region()`. AFAICT, the `body_span` never changed. Why incur the code of recomputing the source_file and file_name every time this is called?

File: compiler/rustc_mir/src/transform/coverage/mod.rs
Comment (richkadel) [inline compiler/rustc_mir/src/transform/coverage/mod.rs:511]: These look like the only pertinent changes in this file. Let me know if I'm missing something.

Very exciting if this works.

File: src/test/run-make-fulldeps/coverage/compiletest-ignore-dir
Comment (richkadel) [inline src/test/run-make-fulldeps/coverage/compiletest-ignore-dir:1]: Thank you for fixing these comments that I forgot to update!!

**Review by richkadel — COMMENTED** [review]
@Swatinem - This `doctest_offset_line()` call looks promising!

I know this is a Draft, but I commented on what you have so far.

I assume you'll want to add a test(s) in `run-make-fulldeps/coverage/`.  Just run `./x.py test src/test/run-make-fulldeps/coverage --bless` once you've written the new test, and it will generate the coverage results, etc.

File: compiler/rustc_mir/src/transform/coverage/mod.rs
Comment (richkadel) [reply compiler/rustc_mir/src/transform/coverage/mod.rs:494]: Ah, I just saw a case (in reproducing the bug mentioned in #79725) where the `span` is in a different file.

I thought I was preventing this, and in many cases it does not happen.

So we may have to do what you're doing here. But I do want to dig into this a bit more in case. I wouldn't be surprised if there are side effects because I made some assumptions that the spans would only come from the body_span.

I'll add a note once I learn more, either way.

Comment (richkadel) [reply compiler/rustc_mir/src/transform/coverage/mod.rs:494]: I did upload a PR (#79818) to ensure `Instrumentor` generates coverage for only one file. I don't know if that fixes a problem you were trying to address here, or something else. Take a look and see if there's an impact either way. Thanks!

File: src/librustdoc/doctest.rs
Comment (Swatinem) [inline src/librustdoc/doctest.rs:765]: This basically reverts half of https://github.com/rust-lang/rust/pull/79413 since https://github.com/rust-lang/cargo/pull/8954 is a proper solution to this. CC @jyn514

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt
Comment (Swatinem) [inline src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt:35]: tbh, I have no idea where these changes are coming from…

File: src/librustdoc/doctest.rs
Comment (jyn514) [reply src/librustdoc/doctest.rs:765]: Sure, seems reasonable. I always want to do things through cargo instead of rustdoc where possible because cargo has more information.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt
Comment (jyn514) [reply src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt:35]: Probably the function name changed. This hash depends on quite a lot of things, if the name changed I wouldn't worry about it.

File: src/librustdoc/doctest.rs
Comment (richkadel) [inline src/librustdoc/doctest.rs:549]: Is `test_id` only non-empty when calling this from the `InstrumentCoverage` pass? Or are you proposing changing the inner function name for all doctests from now on? I'm not sure if that's a good thing, or if it would be better to limit the effect of the change. Maybe that change should not be limited to only when coverage is enabled.

A couple of suggestions here:

1. Instead of sticking with the literal `str` for this value, how about just making this a call to `format!()` and avoid the need to `replace` a pattern later. This also avoids the possibility that the pattern might actually be in the test itself, likely causing a bug.

2. Since the `returns_result == true` case already implements a wrapped function, and you're doing the same, can I suggest keeping as much of the original implementation and style as before, so without coverage enabled, things are the same?

For example, would something like this work?

```rust
let wrapped_fn_name = if test_id.is_empty() {
    "_inner"
} else {
    "_inner_" + test_id
};
let (main_pre, main_post) = if returns_result {
    format!("fn main() { fn {}() -> Result<(), impl core::fmt::Debug> {", wrapped_fn_name),
    format!("}\n{}().unwrap() }", wrapped_fn_name),
} else {
    format!("fn main() { fn {}() {", wrapped_fn_name),
    format!("}\n{}() }", wrapped_fn_name),
};
```

This still does make one change to the test structure: if `test_id` is empty and the `returns_result` is false. I'm not sure it's worth the effort to avoid creating the wrapper only in that case, so unless it causes big issues in downstream tests (hard to believe it would), you probably don't need that special case handling.

File: src/librustdoc/doctest.rs
Comment (richkadel) [reply src/librustdoc/doctest.rs:550]: (I wrote this comment before the previous comment. The sample code (if you think that works) keeps the original newline placement, assuming that's ok.

File: src/librustdoc/doctest.rs
Comment (richkadel) [inline src/librustdoc/doctest.rs:759]: nit: change this `test_id` to `file`

File: src/librustdoc/doctest.rs
Comment (richkadel) [inline src/librustdoc/doctest.rs:757]: I don't know the background here. Some of the changes you are proposing seem to have significant side effects that might affect downstream testing, especially if not limited to "coverage"-enabled tests.

1. Is it OK that `cratename` is no longer part of the `path`?

2. Why change the path at all?

3. And I guess I understand the need to change the `map()` for the `test_id` to check `is_ascii_alphanumeric` since you need to use it as a function name identifier, but again, using that in the path name as well seems like a shortcut with small benefit but a downstream effect that is broader in scope than just fixing doctest coverage.

If other compiler team members are OK with this, that's fine with me, but I want to make sure someone else takes a look.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.uses_crate.txt
Comment (richkadel) [inline src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.uses_crate.txt:22]: This is not unusual for Windows to swap these, and should be ignored by the diff flags.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.async.txt
Comment (richkadel) [inline src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.async.txt:36]: These changes I'm not sure are OK, but I'm not sure why they are changing either. 

I'm not sure if this will break cross-platform tests.

Might require a closer look.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt
Comment (richkadel) [reply src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt:35]: Actually it's not a change, per se... The ordering has been swapped.

This is a problem with `llvm-cov` output. It generates output for multiple instantiations of generics in a non-deterministic order.

If the diff tests are going to fail because of this, we may need to make some changes. 

But I think this should have already been a problem, in which case maybe it's already handled.

I'll take another look at this later.

**Review by richkadel — CHANGES_REQUESTED** [review]
Looks really cool so far. 

I'll take another look at the issues with non-deterministic order in some `llvm-cov` test results. (A known issue that's HOPEFULLY already addressed, but generating the expected test results on Windows will highlight those issues for sure.)

File: src/librustdoc/doctest.rs
Comment (jyn514) [reply src/librustdoc/doctest.rs:757]: Maybe @ollie27 would be a good reviewer? I'm not very familiar with this part of rustdoc.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.async.txt
Comment (richkadel) [reply src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.async.txt:36]: @Swatinem - I think you are safe if the `expected_show_coverage_counters.*` files change by sort order.

It looks like the Makefile does not fail the test if the `_counters` files don't match. There are a couple of reasons why those files aren't stable enough, mostly due to sort order.

**Comment by camelid** [comment]
Looks like you just need to run `./x.py test --bless src/test/run-make-fulldeps/coverage-reports` to fix the conflicts.

File: src/librustdoc/doctest.rs
Comment (Swatinem) [reply src/librustdoc/doctest.rs:549]: > Or are you proposing changing the inner function name for all doctests from now on?

Yes, as `test_id` is always defined with my proposed change. I think its an easy change to make, and I don’t think it should have any other effects, but I could be wrong.

File: src/librustdoc/doctest.rs
Comment (Swatinem) [reply src/librustdoc/doctest.rs:757]: Yes, the change is mostly to be able to use it as part of the identifier in code. The filename is only used for `--persist-doctests`, which is basically generating kind-of "random" filenames for each test. I think most tools using this will just enumerate all the executables in the `--persist-doctests` directory. Also, I recently fixed that part of the code already because it was broken on windows (did not replace `\` properly)

**Comment by Swatinem** [comment]
👍 re-blessed the coverage tests and improved the code wrapping the doctest main, also adding more comments explaining the line offsets and its effect on code coverage.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt
Comment (richkadel) [reply src/test/run-make-fulldeps/coverage-reports/expected_show_coverage_counters.generics.txt:35]: This should be OK. As mentioned in another comment, the tests don't fail if counters files don't match, for multiple reasons.

File: src/librustdoc/doctest.rs
Comment (richkadel) [inline src/librustdoc/doctest.rs:780]: As mentioned in the nit comment (https://github.com/rust-lang/rust/pull/79762/files#r541196922), changing the first `let test_id =` (on line 773) to `let file =` would be more straightforward, and it would also allow you to change this line to just `file,`.

I still recommend that change.

**Comment by richkadel** [comment]
Looks great, except for the one earlier `nit` comment (and my follow-up just now); plus the test failures that seem to be caused by your change somehow. Any ideas?

```
failures:
    [rustdoc] rustdoc/playground-arg.rs
    [rustdoc] rustdoc/playground.rs
    [rustdoc] rustdoc/playground.rs
```

FYI, I don't have authority to approve PRs in the rust-lang/rust repo. It looks like @estebank was recommended. In case @estebank is not available for review, you could request a review (with `r? userid`) from either @tmandry or @wesleywiser.

Thanks for this solution!!

**Comment by jyn514** [comment]
I think this should have a rustdoc reviewer, but I don't know this code well enough.

r? @GuillaumeGomez maybe?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
It definitely looks like a nice improvement! I didn't see anything problematic on rustdoc side (and discovered quite a lot of new test suites while reading this PR haha). Thanks!

**Comment by tmandry** [comment]
@bors delegate+

r=me with rebase (and after #80109 lands you'll want to remove some of the test files, I think)

File: src/librustdoc/doctest.rs
Comment (richkadel) [inline src/librustdoc/doctest.rs:550]: It looks like the test errors (snake case) are due to a double-underscore after `_doctest_main`, which probably means `test_id` can start with an underscore.

Maybe you are converting a path with a leading slash and the slash becomes a leading underscore, or something like that. If it's that simple, you can drop the leading underscore either here or where the `test_id` is generated.

File: src/test/run-make-fulldeps/coverage/doctest.rs
Comment (richkadel) [inline src/test/run-make-fulldeps/coverage/doctest.rs:26]: I just realized, the coverage results in `src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.doctest.txt` don't show any examples of coverage count differences within a line.

I think we need at least one, to make sure the column positions are also computed correctly.

Look, for example, at:

https://github.com/rust-lang/rust/blob/bd2f1cb2785f87177249e2bdb628ed782fcd8def/src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.conditions.txt#L30

```rust

   28|      1|    if countdown > 7 {
   29|      1|        countdown -= 4;
   30|      1|    } else if countdown > 2 {
                            ^0
   31|      0|        if countdown < 1 || countdown > 5 || countdown != 9 {
   32|      0|            countdown = 0;
   33|      0|        }
```

There are other examples, in this sample program and others, showing the column-specific counter position. (It's even more clear when viewing coverage results with `--color`, in a terminal.)

It should be pretty easy to add something similar, and compare the doctest column position to the equivalent in a standard rust source coverage result.

**Review by richkadel — CHANGES_REQUESTED** [review]
Sorry for the last minute comments.

I think I see the problem with the failing tests, so I added a comment suggesting a possible easy fix.

I also noticed something we should have verified through your new test (column positions), which should be easy to add to the test, and necessary. It may require a change to your solution, unless the generated test code is already indented to the same original column positions. Should be verified either way.

Thanks!

File: src/librustdoc/doctest.rs
Comment (Swatinem) [reply src/librustdoc/doctest.rs:550]: Ah yes, i rather tried to add a `#[allow()]` directive in there, I didn’t want to go too crazy sanitizing this.

File: src/test/run-make-fulldeps/coverage/doctest.rs
Comment (Swatinem) [reply src/test/run-make-fulldeps/coverage/doctest.rs:26]: As the PR description says, the columns are not yet correct. I think a solution for that could be a lot more complex, as rustdoc can have non-uniform column offsets, such as:

````
/// ```
/// some_normal_code();
/// # executed_but_hidden_code();
/// ```
````

In the intermediate generated code, these have the same (no) indentation.

Another interesting case is that you can mix `///`-style comments, and doc attributes, with a lot more complex cases that I actually first read about here: https://blog.guillaume-gomez.fr/articles/2020-11-11+New+doc+comment+handling+in+rustdoc

So in those cases, the column offsets can differ for each line.

Comment (richkadel) [reply src/test/run-make-fulldeps/coverage/doctest.rs:26]: Ah. I miss that. That's fine then. It certainly a big improvement with the corrected line numbers and we can save the column positions for a future PR. Thanks!

**Comment by Swatinem** [comment]
So I think this is finally ready to land, lets do this…

@bors r+

**Comment by richkadel** [comment]
We'll need to look at the make file changes. Mac OS make, and bash, have some limitations, and even bugs. Recently I ran into one where substitutions actually produced code that was not in the original makefile. Having lots of lines of script with backslashes at new lines seem to cause that last bug. But it's hard to test because it doesn't happen on all variants. One rust developer reproduced it with xcode 12, but I have xcode 11. That may not be the issue here though.

**Comment by Swatinem** [comment]
It looks like macos still has the deprecated python2 by default 🤦‍♂️ 
Lets try again with the explicit new `print` statement import…

@bors r+

**Comment by jyn514** [comment]
@Swatinem in the future, please use `r=richkadel`, you generally aren't supposed to approve your own PR :P not a big deal though.,

**Comment by Swatinem** [comment]
Absolutely, I thought the "delegate" meant me. anyhow ;-)

**Comment by Swatinem** [comment]
Ah yes, https://github.com/rust-lang/rust/pull/79762#issuecomment-747809695 says

> ✌️ @Swatinem can now approve this pull request

So that confused me :-D

**Comment by richkadel** [comment]
Actually it should be r=tmandry I think. I don't have approval authority.

**Pull request merged**
